### PR TITLE
VR-4346 VR-4347 initialize multipart, get URL

### PR DESCRIPTION
**Pull request by evgeniilezhnin**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/nfs/NFSService.java [hunk 1]
```diff
@@ -243,4 +243,9 @@ public String generatePresignedUrl(String artifactPath, String method) {
       throw StatusProto.toStatusRuntimeException(status);
     }
   }
+
+  @Override
+  public void initializeMultipart(String s3Key) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/nfs/NFSService.java]: We will not be supporting it in NFS for the time being.

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java [hunk 2]
```diff
@@ -1362,6 +1362,7 @@ public void logArtifact(
       }
       Artifact artifact = artifacts.get(0);
 
+      artifactStoreDAO.initializeMultipart(artifact.getKey());
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java]: Why are you calling this here? The client will explicitly initialize via the new API on getUrlForArtifact. I don't think we have to or should do it before.

Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java]: I think because the task says to add it to log artifact: https://vertaai.atlassian.net/browse/VR-4346 cc: @convoliution

Comment (convoliution) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java]: My proposal puts it in `logArtifact`, yes. 
If it's recommended to put it in `getUrlForArtifact` @conradoverta , then I can revise the document and tickets in the morning.

Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java]: Not every artifact will be uploaded, so it doesn't seem like a good idea. I'd also recommend against creating a new flow if we don't have to. We can always revisit, but let's focus on delivering one thing at a time.

File: backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/nfs/NFSService.java (hunk 1 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/artifactStore/storageservice/nfs/NFSService.java]: changed that it will return null that will mean failed precondition

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java (hunk 2 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java]: moved initialization to getUrlForArtifact

**Review by convoliution — APPROVED** [review]

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java [hunk 3]
```diff
@@ -2247,4 +2250,55 @@ public void deleteLogVersionedInputs(Session session, Long repoId, String commit
...
+          (artifactEntityMap != null && artifactEntityMap.containsKey(ModelDBConstants.ARTIFACTS))
+              ? artifactEntityMap.get(ModelDBConstants.ARTIFACTS)
+              : Collections.emptyList();
+      for (ArtifactEntity artifactEntity : result) {
+        if (artifactEntity.getKey().equals(key)) {
+          String uploadId;
+          if (partNumber != 0) {
+            uploadId = artifactEntity.getUploadId();
+            if (uploadId == null) {
+              uploadId = initializeMultipart.apply(key).orElse(null);
```
Comment (convoliution) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: `initializeMultipart()` needs the S3 key (which is `path` from the artifact), not the ModelDB artifact key!
```suggestion
              uploadId = initializeMultipart.apply(artifactEntity.getPath()).orElse(null);
```
This is why we get `NoSuchUpload` from S3 when we try to use the presigned URL:
The key used to generate the URL was different from the one used for `initializeMultipart()`.

I have verified that this change fixes the `NoSuchUpload` error.
If `initializeMultipart()` is called anywhere else, make sure those are correct as well!

**Review by convoliution — CHANGES_REQUESTED** [review]

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java (hunk 3 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: Thank you very much! That works

**Review by convoliution — COMMENTED** [review]
I have confirmed functionality. Adding @ravishetye as reviewer for code.

**Review by convoliution — APPROVED** [review]
Oops forgot to approve
@evgeniilezhnin Don't merge until Ravi has a chance to look over the code!

File: protos/protos/public/modeldb/CommonService.proto [hunk 4]
```diff
@@ -113,6 +113,7 @@ message GetUrlForArtifact {
     message Response {
         string url = 1;
         map<string, string> fields = 2;
+        bool multipart_upload_ok = 3;
```
Comment (conradoverta) [inline protos/protos/public/modeldb/CommonService.proto]: @convoliution are these changes intentional? Asking because it's different from your PR that changed the protos.

File: backend/src/main/java/ai/verta/modeldb/entities/ArtifactPartEntity.java [hunk 5]
```diff
@@ -44,9 +43,9 @@ public boolean equals(Object o) {
       return false;
     }
     ArtifactPartEntity that = (ArtifactPartEntity) o;
-    return partNumber == that.partNumber &&
-        Objects.equals(artifactEntity, that.artifactEntity) &&
-        Objects.equals(etag, that.etag);
+    return partNumber == that.partNumber
+        && Objects.equals(artifactEntity, that.artifactEntity)
+        && Objects.equals(etag, that.etag);
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/entities/ArtifactPartEntity.java]: Avoid placing formatting changes with other changes.

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java [hunk 6]
```diff
@@ -2247,4 +2250,55 @@ public void deleteLogVersionedInputs(Session session, Long repoId, String commit
...
+              : Collections.emptyList();
+      for (ArtifactEntity artifactEntity : result) {
+        if (artifactEntity.getKey().equals(key)) {
+          String uploadId;
+          if (partNumber != 0) {
+            uploadId = artifactEntity.getUploadId();
+            if (uploadId == null) {
+              uploadId = initializeMultipart.apply(artifactEntity.getPath()).orElse(null);
+            }
+            if (uploadId == null) {
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: What value will `uploadId` be at this moment for NFS?

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java [hunk 7]
```diff
@@ -2247,4 +2250,55 @@ public void deleteLogVersionedInputs(Session session, Long repoId, String commit
           .build();
     }
   }
+
+  @Override
+  public Entry<String, String> getExperimentRunArtifactsS3PathAndMultipartUploadID(
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: I think you meant `getExperimentRunArtifactS3PathAndMultipartUploadID`, not `artifacts`, right? It's returning a single value.

**Review by convoliution — APPROVED** [review]

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java [hunk 8]
```diff
@@ -1222,27 +1224,36 @@ public void getUrlForArtifact(
...
         errorMessage =
             "Code versioning artifact not found at experimentRun, experiment and project level";
         s3Key = getUrlForCode(request);
+        uploadId = null;
       } else if (request.getArtifactType() == ArtifactType.DATA) {
         errorMessage = "Data versioning artifact not found";
-        s3Key = getUrlForData(request);
+        Entry<String, String> s3KeyUploadId = getUrlForData(request);
+        s3Key = s3KeyUploadId.getKey();
+        uploadId = s3KeyUploadId.getValue();
```
Comment (conradoverta) [inline backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java:1244]: @convoliution didn't we deprecate data as artifact?

**Review by conradoverta — COMMENTED** [review]
Looking good, just some clarification.

File: protos/protos/public/modeldb/CommonService.proto (hunk 4 above)
Comment (convoliution) [reply protos/protos/public/modeldb/CommonService.proto:116]: The protos changes in this diff are actually already in my protos PR. I guess GitHub isn't showing an accurate diff because these backend PRs are chained on top of each other.

@evgeniilezhnin Could you revert the protos changes in this PR?

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java (hunk 8 above)
Comment (convoliution) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunServiceImpl.java:1244]: I can't recall whether we decided to deprecate it, but maybe not:
The Client [still has the logic for uploading data as an artifact](https://github.com/VertaAI/modeldb/blob/095f2bd/client/verta/verta/client.py#L2676) (if the user doesn't use `DatasetVersion`),
and past-me didn't mark it as deprecated.

File: protos/protos/public/modeldb/CommonService.proto (hunk 4 above)
Comment (convoliution) [reply protos/protos/public/modeldb/CommonService.proto]: The protos changes in this diff are actually already in my protos PR. I guess GitHub isn't showing an accurate diff because these backend PRs are chained on top of each other.

@evgeniilezhnin Could you revert the protos changes in this PR?

**Review by convoliution — CHANGES_REQUESTED** [review]

File: protos/protos/public/modeldb/CommonService.proto (hunk 4 above)
Comment (evgeniilezhnin) [reply protos/protos/public/modeldb/CommonService.proto]: merged the base branch, the changes are gone

File: backend/src/main/java/ai/verta/modeldb/entities/ArtifactPartEntity.java (hunk 5 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/entities/ArtifactPartEntity.java]: removed it

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java (hunk 6 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: null. But Michael said that I should remove this validation check here https://github.com/VertaAI/modeldb/pull/660#discussion_r431373980 so I'm removing it here too

File: backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java (hunk 7 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java]: fixed

**Review by convoliution — APPROVED** [review]

**Review by conradoverta — APPROVED** [review]
